pallets/gear/src/internal.rs

```diff
@@ -634,22 +698,42 @@ where
                     Self::remove_gas_reservation_with_task(dispatch.source(), reservation_id);
                 }
             }
-        }
 
-        if !dispatch.value().is_zero() {
-            // Reserving value from source for future transfer or unreserve.
-            CurrencyOf::<T>::reserve(&from, value)
-                .unwrap_or_else(|e| unreachable!("Unable to reserve requested value {:?}", e));
+            // Calculating correct hold bound to lock gas
+            let maximal_hold = hold_builder.clone().maximum_for_message(dispatch.id());
+            let bn_delay = delay.saturated_into::<BlockNumberFor<T>>();
+            let hold = hold_builder.duration(bn_delay).min(maximal_hold);
+
+            // Locking funds for holding.
+            GasHandlerOf::<T>::lock(dispatch.id(), hold.lock())
+                .unwrap_or_else(|e| unreachable!("GasTree corrupted! {:?}", e));
+
+            if hold.expected_duration().is_zero() {
+                // TODO: Replace with unreachable call after:
+                // - `HoldBound` safety usage stabilized;
+                log::error!("Failed to figure out correct wait hold bound");
+                return;
+            }
+
+            if !dispatch.value().is_zero() {
+                // Reserving value from source for future transfer or unreserve.
+                CurrencyOf::<T>::reserve(&from, value)
+                    .unwrap_or_else(|e| unreachable!("Unable to reserve requested value {:?}", e));
+            }
         }
 
         // Saving id to allow moving dispatch further.
         let message_id = dispatch.id();
 
+        // Add block number of insertation
+        let mut stored_dispatch = dispatch.into_stored();
+        stored_dispatch.set_stored_block_number(Self::block_number().unique_saturated_into());
```

Comment:
that's always bad approach cause at least after into_stored_conversion your inner block number is zeroed for no reason
but it will anyway be remove once moved into tuple as storage map value